pydantic/schema.py

```diff
@@ -671,13 +677,18 @@ def field_singleton_schema(  # noqa: C901 (ignore complexity)
     for type_, t_schema in field_class_to_schema_enum_disabled:
         if issubclass(field.type_, type_):
             return t_schema, definitions
-    if issubclass(field.type_, main.BaseModel):
+    # Handle dataclass-based models
+    field_type = field.type_
+    if lenient_issubclass(getattr(field_type, '__pydantic_model__', None), main.BaseModel):
+        field_type = cast(Type['dataclasses.DataclassType'], field_type)
```

Comment:
Sorry for further comments, should have said this before.

Better approach than `cast`ing I think:

```py
field_type: Union[Type['main.BaseModel'], Type['dataclasses.DataclassType']] = field.type_
if lenient_issubclass(getattr(field_type, '__pydantic_model__', None), main.BaseModel):
    field_type = field_type.__pydantic_model__
```

Reply:
Union type in mypy sometimes is tricky, it requires the existence of `__pydantic_model__` attribute on both `Type['main.BaseModel']` and `Type['dataclasses.DataclassType']`.

For the above code, it will report the following error,
```
pydantic\schema.py:683: error: Item "MetaModel" of "Union[Type[BaseModel], Type[DataclassType]]" has no attribute "__pydantic_model__"
make: *** [mypy] Error 1
```

Reply:
ok